File: solutions/Python/validate_binary_search_tree_inorder.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def isValidBST(root: Optional[TreeNode]) -> bool:
    prev = [float('-inf')]
    def dfs(node: Optional[TreeNode]) -> bool:
        if not node: return True
        if not dfs(node.left): return False
        if node.val <= prev[0]: return False
        prev[0] = node.val
        return dfs(node.right)
    return dfs(root)

if __name__ == "__main__":
    root = TreeNode(2)
    root.left, root.right = TreeNode(1), TreeNode(3)
    print(isValidBST(root))  # True

File: solutions/Golang/validate_binary_search_tree_inorder.go
package main

import "math"

type TreeNode struct { Val int; Left, Right *TreeNode }

func IsValidBST(root *TreeNode) bool {
    prev := []float64{math.Inf(-1)}
    var dfs func(*TreeNode) bool
    dfs = func(n *TreeNode) bool {
        if n == nil { return true }
        if !dfs(n.Left) { return false }
        if float64(n.Val) <= prev[0] { return false }
        prev[0] = float64(n.Val)
        return dfs(n.Right)
    }
    return dfs(root)
}
